torch/_higher_order_ops/scan.py

```diff
@@ -43,8 +104,64 @@ def wrap_combine_fn_flat(
     return [*carry_flat, *combined_flat]
 
 
-def _extract_carry_and_out(flat_out: List[Any], num_carry: int):
-    return flat_out[:num_carry], flat_out[num_carry:]
+def create_fw_bw_graph_combinefn(combine_fn, init, xs, dim, additional_inputs):
+    # See Note [HOP create fw_bw graph] in create_fw_bw_graph in utils.py
+
+    # Helper wrapper for the autograd forward.
+    # This wrapper ensures that the forward returns all carries
+    # instead of only the last one
+    # The gradients of the carries forwarded to the output are
+    # detached in order not to raise problems with the function aliasing outputs
+    def wrapper_combine_fn(*args):
+        new_carry, y = _extract_carry_and_out(combine_fn(*args), len(init))
+        return [*new_carry, *[n_c.clone().detach() for n_c in new_carry], *y]
+
+    with suspend_functionalization(), disable_functional_mode():
+        with disable_proxy_modes_tracing():
+            num_init = len(init)
+
+            fw_init = [pytree.tree_map(_from_fun, x) for x in init]
+            fw_xs = [pytree.tree_map(_from_fun, x).select(dim, 0) for x in xs]
+
+            carry, outs = _extract_carry_and_out(
+                wrapper_combine_fn(*fw_init, *fw_xs, *additional_inputs), num_init
+            )
+            fw_carry, fw_outputs = [pytree.tree_map(_from_fun, c) for c in carry], [
+                pytree.tree_map(_from_fun, o) for o in outs
+            ]
+            if any(carry.shape != ini.shape for carry, ini in zip(fw_carry, init)):
+                raise RuntimeError(
+                    "Expect carry produced by combine_fn to only contains tensors. "
+                    f"Got types {[type(carry) for carry in fw_carry]}."
+                )
+            if any(not isinstance(carry, torch.Tensor) for carry in fw_carry):
+                raise RuntimeError(
+                    "Expect carry produced by combine_fn to only contains tensors. "
+                    f"Got types {[type(carry) for carry in fw_carry]}."
+                )
+            if any(not isinstance(out, torch.Tensor) for out in fw_outputs):
+                raise RuntimeError(
+                    "Expect outputs produced by combine_fn to only contains tensors. "
+                    f"Got types {[type(out) for out in fw_outputs]}."
+                )
+
+            # TODO: There is a major issue that the create_fw_bw in the higher_order_op is invoked twice:
```

Comment:
We can probably move to where we do the torch.is_grad_enabled() check (need to change the wording a bit).

Reply:
Done